src/katgpucbf/xbgpu/engine.py

```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+            dump_interval_s=self.dump_interval_s,
+            send_rate_factor=engine.send_rate_factor,
+            channel_offset=engine.channel_offset_value,  # Arbitrary for now - depends on F-Engine stream
+            context=context,
+            packet_payload=engine.dst_packet_payload,
+            stream_factory=lambda stream_config, buffers: make_stream(
+                output_name=output.name,
+                dest_ip=output.dst.host,
+                dest_port=output.dst.port,
+                interface_ip=engine.dst_interface,
+                ttl=engine.dst_ttl,
+                use_ibv=engine.dst_ibv,
+                affinity=engine.dst_affinity,
+                comp_vector=engine.dst_comp_vector,
+                stream_config=stream_config,
+                buffers=buffers,
+            ),
+            tx_enabled=init_tx_enabled,
+        )
+
+        self._populate_sensors()
+
+    def _populate_sensors(self) -> None:  # noqa: D102
+        sensors = self.engine.sensors
+        # Static sensors
+        sensors.add(
+            aiokatcp.Sensor(
+                str,
+                f"{self.output.name}.chan-range",
+                "The range of channels processed by this X-engine, inclusive",
+                default=f"({self.engine.channel_offset_value},"
+                f"{self.engine.channel_offset_value + self.engine.n_channels_per_stream - 1})",
+                initial_status=aiokatcp.Sensor.Status.NOMINAL,
+            )
+        )
+        # Dynamic sensors
+        sensors.add(
+            aiokatcp.Sensor(
+                bool,
+                f"{self.output.name}.rx.synchronised",
+                "For the latest accumulation, was data present from all F-Engines.",
+                default=False,
+                initial_status=aiokatcp.Sensor.Status.ERROR,
+                status_func=lambda value: aiokatcp.Sensor.Status.NOMINAL if value else aiokatcp.Sensor.Status.ERROR,
+            )
+        )
+        sensors.add(
+            aiokatcp.Sensor(
+                int,
+                f"{self.output.name}.xeng-clip-cnt",
+                "Number of visibilities that saturated",
+                default=0,
+                initial_status=aiokatcp.Sensor.Status.NOMINAL,
+            )
+        )
+
+    async def _flush_accumulation(self, tx_item: TxQueueItem, next_accum: int) -> TxQueueItem:
+        """Emit the current `tx_item` and prepare a new one."""
+        if tx_item.batches == 0:
+            # We never actually started this accumulation. We can just
+            # update the timestamp and continue using it.
+            tx_item.timestamp = next_accum * self.timestamp_increment_per_accumulation
+            return tx_item
+
+        # present_ants only takes into account batches that have
+        # been seen. If some batches went missing entirely, the
+        # whole accumulation is bad.
+        if tx_item.batches != self.output.heap_accumulation_threshold:  # type: ignore
```

Comment:
I added it here because mypy complained
> "Output" has no attribute "heap_accumulation_threshold"

Reply:
That's probably because `self.output` is defined in the base class as `Output`, so it doesn't know that it's actually an XOutput. That problem should vanish once you remove `output` from the base class constructor and handle it in XPipeline.